flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/ExecutingTest.java

```diff
@@ -0,0 +1,451 @@
+            Executing exec = getExecutingState(ctx);
+            exec.onEnter();
+            ctx.setExpectFailing(
+                    (failingArguments -> {
+                        assertThat(failingArguments.getExecutionGraph(), notNullValue());
+                        assertThat(failingArguments.getFailureCause().getMessage(), is(failureMsg));
+                    }));
+            ctx.setHowToHandleFailure(Executing.FailureResult::canNotRestart);
+            exec.handleGlobalFailure(new RuntimeException(failureMsg));
+        }
+    }
+
+    @Test
+    public void testTransitionToRestarting() throws Exception {
+        final Duration duration = Duration.ZERO;
+        try (MockExecutingContext ctx = new MockExecutingContext()) {
+            Executing exec = getExecutingState(ctx);
+            exec.onEnter();
+            ctx.setExpectRestarting(
+                    (restartingArguments ->
+                            assertThat(restartingArguments.getBackoffTime(), is(duration))));
+            ctx.setHowToHandleFailure((t) -> Executing.FailureResult.canRestart(duration));
+            exec.handleGlobalFailure(new RuntimeException("Recoverable error"));
+        }
+    }
+
+    @Test
+    public void testTransitionToCancelling() throws Exception {
+        try (MockExecutingContext ctx = new MockExecutingContext()) {
+            Executing exec = getExecutingState(ctx);
+            exec.onEnter();
+            ctx.setExpectCancelling(assertNonNull());
+            exec.cancel();
+        }
+    }
+
+    @Test
+    public void testTransitionToFinishedOnTerminalState() throws Exception {
+        try (MockExecutingContext ctx = new MockExecutingContext()) {
+            Executing exec = getExecutingState(ctx);
+            exec.onEnter();
+            ctx.setExpectFinished(
+                    archivedExecutionGraph ->
+                            assertThat(archivedExecutionGraph.getState(), is(JobStatus.FAILED)));
+
+            ctx.setExpectedStateChecker((state) -> state == exec);
+            // transition EG into terminal state, which will notify the Executing state about the
+            // failure (async via the supplied executor)
+            exec.getExecutionGraph().failJob(new RuntimeException("test failure"));
+        }
+    }
+
+    @Test
+    public void testTransitionToFinishedOnSuspend() throws Exception {
+        try (MockExecutingContext ctx = new MockExecutingContext()) {
+            Executing exec = getExecutingState(ctx);
+            ctx.setExpectedStateChecker((state) -> state == exec);
+
+            ctx.setExpectFinished(
+                    archivedExecutionGraph -> {
+                        assertThat(archivedExecutionGraph.getState(), is(JobStatus.SUSPENDED));
+                    });
+            exec.onEnter();
+            exec.suspend(new RuntimeException("suspend"));
+        }
+    }
+
+    @Test
+    public void testScaleUp() throws Exception {
```

Comment:
It is always a good idea to use a test function name to express what you wanna test and what the expected outcome is. Here, for example, it could be `testNotifyNewResourcesAvailableWithCanScaleUpTransitionsToRestarting`.